zkevm-circuits/src/evm_circuit/witness.rs

```diff
@@ -0,0 +1,520 @@
+            return ExecutionState::DUP;
+        }
+        if step.op.is_push() {
+            return ExecutionState::PUSH;
+        }
+        if step.op.is_swap() {
+            return ExecutionState::SWAP;
+        }
+        match step.op {
+            OpcodeId::ADD => ExecutionState::ADD,
+            OpcodeId::SUB => ExecutionState::ADD,
+            OpcodeId::EQ => ExecutionState::CMP,
+            OpcodeId::GT => ExecutionState::CMP,
+            OpcodeId::LT => ExecutionState::CMP,
+            OpcodeId::SIGNEXTEND => ExecutionState::SIGNEXTEND,
+            OpcodeId::STOP => ExecutionState::STOP,
+            OpcodeId::AND => ExecutionState::BITWISE,
+            OpcodeId::XOR => ExecutionState::BITWISE,
+            OpcodeId::OR => ExecutionState::BITWISE,
+            OpcodeId::POP => ExecutionState::POP,
+            OpcodeId::PUSH32 => ExecutionState::PUSH,
+            OpcodeId::BYTE => ExecutionState::BYTE,
+            OpcodeId::MLOAD => ExecutionState::MEMORY,
+            OpcodeId::MSTORE => ExecutionState::MEMORY,
+            OpcodeId::MSTORE8 => ExecutionState::MEMORY,
+            OpcodeId::JUMPDEST => ExecutionState::JUMPDEST,
+            OpcodeId::PC => ExecutionState::PC,
+            _ => unimplemented!("invalid opcode {:?}", step.op),
+        }
+    }
+}
+
+impl From<&bus_mapping::bytecode::Bytecode> for Bytecode {
+    fn from(b: &bus_mapping::bytecode::Bytecode) -> Self {
+        Bytecode::new(b.to_bytes())
+    }
+}
+
+fn step_convert(
+    prev: Option<&bus_mapping::circuit_input_builder::ExecStep>,
+    step: &bus_mapping::circuit_input_builder::ExecStep,
+    ops_len: (usize, usize, usize),
+) -> ExecStep {
+    let (stack_ops_len, memory_ops_len, _storage_ops_len) = ops_len;
+    //println!("prev is {:#?}", prev);
+    let result = ExecStep {
+        rw_indices: step
+            .bus_mapping_instance
+            .iter()
+            .map(|x| {
+                let index = x.as_usize() - 1;
+                match x.target() {
+                    bus_mapping::operation::Target::Stack => index,
+                    bus_mapping::operation::Target::Memory => {
+                        index + stack_ops_len
+                    }
+                    bus_mapping::operation::Target::Storage => {
+                        index + stack_ops_len + memory_ops_len
+                    }
+                }
+            })
+            .collect(),
+        execution_state: ExecutionState::from(step),
+        rw_counter: usize::from(step.gc),
+        program_counter: usize::from(step.pc) as u64,
+        stack_pointer: 1024 - step.stack_size,
+        gas_left: step.gas_left.0,
+        gas_cost: step.gas_cost.as_u64(),
+        opcode: Some(step.op),
+        // geth increases memory size before making trace,
+        // so the memory size in ExecStep is not what we expect to see.
+        // We have to use the memory size of previous step as correct memory
+        // size before each step
```

Comment:
Add link to geth: https://github.com/ethereum/go-ethereum/blob/master/core/vm/interpreter.go#L224-L233 in comments

Reply:
addressed in https://github.com/scroll-tech/zkevm-circuits/pull/30/commits/3184aec2ffa9395a6b2af729c76249dd65d0a899